day3/schematic.go

```diff
@@ -0,0 +1,228 @@
+			for k := 0; k < len(partNumbers[i+1]); k++ {
+				if partNumbers[i+1][k].IsValid {
+					continue
+				}
+				if partNumbers[i+1][k].StartIdx <= j-1 && j-1 <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				} else if partNumbers[i+1][k].StartIdx <= j && j <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				} else if partNumbers[i+1][k].StartIdx <= j+1 && j+1 <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				}
+			}
+		}
+	}
+
+	return partNumbers, nil
+}
+
+// MakeAsteriskMap returns a map representing the surrounding area of an asterisk in a grid.
+// The asterisk is located at the specified line and linePosition.
+// The map contains boolean values indicating whether each position needs to be assessed
+// (i.e. 'false' indicates that this location can be skipped).
+func (schematic *Schematic) MakeAsteriskMap(line int, linePosition int) map[int]map[int]bool {
+	asteriskMap := map[int]map[int]bool{
+		line - 1: {
+			linePosition - 1: true,
+			linePosition:     true,
+			linePosition + 1: true,
+		},
+		line: {
+			linePosition - 1: true,
+			linePosition:     false, // this is the location of the asterisk
+			linePosition + 1: true,
+		},
+		line + 1: {
+			linePosition - 1: true,
+			linePosition:     true,
+			linePosition + 1: true,
+		},
+	}
+	// if on top line
+	if line == 0 {
+		for k := range asteriskMap[line-1] {
+			asteriskMap[line-1][k] = false
+		}
+	}
+	// if on right edge
+	if linePosition == len(schematic.Data[line])-1 {
+		for k := range asteriskMap {
+			asteriskMap[k][linePosition+1] = false
+		}
+	}
+	// if on bottom line
+	if line == len(schematic.Data)-1 {
+		for k := range asteriskMap[line+1] {
+			asteriskMap[line+1][k] = false
+		}
+	}
+	// if on left edge
+	if linePosition == 0 {
+		for k := range asteriskMap {
+			asteriskMap[k][linePosition-1] = false
+		}
+	}
+	return asteriskMap
+}
+
+// SumAllGearRatios performs the calculation described in the Day 3 Part 2
+// problem description.
+func (schematic *Schematic) SumAllGearRatios() (int, error) {
+	gearRatioSum := 0
+	for i := 0; i < len(schematic.Data); i++ {
+		line := schematic.Data[i]
+		for j := 0; j < len(line); j++ {
+			ch := line[j]
+			if ch != '*' {
+				continue
+			}
+			astMap := schematic.MakeAsteriskMap(i, j)
+			parts, err := schematic.FindPartsAroundAsterisk(&astMap)
+			if err != nil {
+				return 0, err
+			}
+			if len(parts) == 2 {
+				gearRatioSum += parts[0] * parts[1]
+			}
+		}
+	}
+	return gearRatioSum, nil
+}
+
+// FindPartsAroundAsterisk walks around the provided asterisk map, finding all part numbers
+// which coincide with the area immediately surrounding an asterisk symbol. It implements
+// memoization so that each character surrounding the asterisk symbol is analyzed only once,
+// simultaneously preventing duplicate parts from being detected.
+func (schematic *Schematic) FindPartsAroundAsterisk(astMap *map[int]map[int]bool) ([]int, error) {
```

Comment:
We actually don't need to send over a pointer to the asterisk map here, we can just send a copy.